compiler/rustc_interface/src/passes.rs

```diff
@@ -866,6 +867,99 @@ fn analysis(tcx: TyCtxt<'_>, (): ()) -> Result<()> {
+            // over-estimate how many vtable entries there are.
+            //
+            // Similarly this is wrong for traits that have methods with possibly-impossible bounds.
+            // For example:
+            //
+            //   trait B<T> { fn f(&self) where T: Copy; }
+            //
+            // Here `dyn B<u8>` will have 4 entries, while `dyn B<String>` will only have 3.
+            // However, since we don't know `T`, we can't know if `T: Copy` holds or not,
+            // thus we lean on the bigger side and say it has 4 entries.
+            traits::vtable::prepare_vtable_segments(tcx, trait_ref, |segment| {
+                match segment {
+                    traits::vtable::VtblSegment::MetadataDSA => {
+                        // If this is the first dsa, it would be included either way,
+                        // otherwise it's needed for upcasting
+                        if std::mem::take(&mut first_dsa) {
+                            unupcasted_cost += 3;
+                        } else {
+                            upcast_cost += 3;
+                        }
+                    }
+
+                    traits::vtable::VtblSegment::TraitOwnEntries { trait_ref, emit_vptr } => {
+                        let existential_trait_ref = trait_ref.map_bound(|trait_ref| {
+                            ty::ExistentialTraitRef::erase_self_ty(tcx, trait_ref)
+                        });
+
+                        // Lookup the shape of vtable for the trait.
+                        let own_existential_entries =
+                            tcx.own_existential_vtable_entries(existential_trait_ref.def_id());
+
+                        let own_entries = own_existential_entries.iter().copied().map(|_def_id| {
+                            // The original code here ignores the method if its predicates are impossible.
+                            // We can't really do that as, for example, all not trivial bounds on generic
+                            // parameters are impossible (since we don't know the parameters...),
+                            // see the comment above.
+
+                            1
+                        });
+
+                        unupcasted_cost += own_entries.sum::<usize>();
+
+                        if emit_vptr {
+                            upcast_cost += 1;
+                        }
+                    }
+                }
+
+                std::ops::ControlFlow::Continue::<std::convert::Infallible>(())
+            });
+
+            sess.code_stats.record_vtable_size(
+                tr,
+                &name,
+                VTableSizeInfo {
+                    trait_name: name.clone(),
+                    size_words_without_upcasting: unupcasted_cost,
+                    size_words_with_upcasting: unupcasted_cost + upcast_cost,
+                    difference_words: upcast_cost,
+                    difference_percent: upcast_cost as f64 / unupcasted_cost as f64 * 100.,
```

Comment:
why isnt this `(size_words_without_upcasting + size_words_with_upcasting) / size_words_without_upcasting`... not sure if i totally understand the math here, that could use some explanation i think :sweat_smile:

Reply:
nvm, I think I get it. This creates a "how much % larger is the new vtable compared to the old table". could use doc tho